SNLS_NLSPowell.h

```diff
@@ -0,0 +1,729 @@
+            // Store the direction of p and x + p
+            double pnorm = 0.0;
+            for (int i = 0; i < m_nDim; i++) {
+                work_arr1[i] *= -1.0; 
+                // If our solution isn't satisfactory later on we'll
+                // reject the solution and calculate the new value of things
+                m_x[i] += work_arr1[i];
+                pnorm += (diag[i] * work_arr1[i]) * (diag[i] * work_arr1[i]);
+            }
+            pnorm = sqrt(pnorm);
+
+            // std::cout << "work_arr1" << std::endl;
+            // snls::linalg::printVec<m_nDim>(work_arr1);
+            // std::cout << "m_x" << std::endl;
+            // snls::linalg::printVec<m_nDim>(m_x);
+
+
+            // On the first iteration adjust the initial step bound
+            if (m_nIters == 1) {
+                m_delta = fmin(m_delta, pnorm);
+            }
+
+            // Evaluate the function at x + p and calculate its norm
+            // Previously iterations kept x and residual seperated by
+            // making use of working arrays...
+            // This is a waste of memory as we can just use residual and x arrays.
+            // Also, we currently don't check on the success of this...
+            // and we should at this back in at some point
+            // However in the mean time, this is to quite the compiler
+            // bool rjSuccess = 
+            this->computeRJ(residual, nullptr);
+            // Need to think about what to do if this fails as things
+            // are handled differently from the TrDL solver.
+            // if ( !(rjSuccess) ) {
+            //     m_status = initEvalFailure;
+            //     return m_status;
+            // }
+            m_fevals += 1;
+
+            const double res1 = snls::linalg::norm<m_nDim>(residual);
+
+            // Compute the scaled actual reduction
+            double actual_reduction = -1.0;
+            // Compute 2nd power
+            if (res1 < m_res) {
+                const double tmp = res1 / m_res;
+                actual_reduction = 1.0 - (tmp * tmp); 
+            }
+
+            // Compute the predicted reduction
+            snls::linalg::matUTriVecMult<m_nDim, m_nDim>(Jacobian, work_arr1, work_arr3);
+            double sum = 0.0;
+            for (int i = 0; i < m_nDim; i++) {
+                // work_arr3 = R \Delta x + Q^T F
+                // where F here is our residual vector
+                work_arr3[i] += qtf[i];
+                sum += (work_arr3[i] * work_arr3[i]);
+            }
+            sum = sqrt(sum);
+
+            double predicted_reduction = 0.0;
+
+            // Compute 2nd power
+            if (sum < m_res) {
+                sum = sum / m_res;
+                predicted_reduction = 1.0 - sum * sum;
+            }
+
+            // Compute the ratio of the actual to the predicted reduction
+            const double ratio = (predicted_reduction > 0.0) ? (actual_reduction / predicted_reduction) : 0.0;
```

Comment:
If m_os != nullptr, the value of this ratio might be something to print. Or at least in the TrDLDenseG case it can be helpful -- but there one is always using the Jacobian from an evaluation, so maybe things are not as straightforward to interpret here. In any case, your output to m_os is rather sparse if someone does want to use that output to help diagnose problems.

Reply:
I can add that information as an output if m_os is not null and we're in a debug build. I imagine another useful spot would be around the dogleg update potentially with the delta X increment and then after the rank1 update of the Q, R, and Q^T f values.

Reply:
The refactor has eliminated this portion of code and now if the user has provided m_os then this sort of thing will be printed. In fact, I'm pretty sure we print all the same information as the `TrDLDenseG` solver now thanks to the fact we now share so much code.